README.md
# go-webrtc-piping
[![CI](https://github.com/nwtgck/go-webrtc-piping/actions/workflows/ci.yml/badge.svg)](https://github.com/nwtgck/go-webrtc-piping/actions/workflows/ci.yml)

WebRTC P2P tunneling/duplex with [Piping Server](https://github.com/nwtgck/piping-server) WebRTC signaling

## Install for Windows
[Download](https://github.com/nwtgck/go-webrtc-piping/releases/download/v0.3.0/webrtc-piping-0.3.0-windows-amd64.zip)

## Install for macOS
```bash
brew install nwtgck/webrtc-piping/webrtc-piping
```

## Install for Ubuntu
```bash
wget https://github.com/nwtgck/go-webrtc-piping/releases/download/v0.3.0/webrtc-piping-0.3.0-linux-amd64.deb
sudo dpkg -i webrtc-piping-0.3.0-linux-amd64.deb 
```

Get more executables in the [releases](https://github.com/nwtgck/go-webrtc-piping/releases).

## TCP tunneling

The following command forwards 8888 port to 9999 port.   

```bash
webrtc-piping tunnel 8888 mypath 
```

```bash
webrtc-piping tunnel -l 9999 mypath
```

## UDP tunneling

Adding -u or --udp option forwards UDP port.

```bash
webrtc-piping tunnel -u 8888 mypath 
```

```bash
webrtc-piping tunnel -ul 9999 mypath
```

## Full-duplex

```bash
echo hello1 | webrtc-piping duplex mypath1 mypath2
# => hello2
```

```bash
echo hello2 | webrtc-piping duplex mypath2 mypath1
# => hello1
```

## Without ICE servers

Specify `--ice-servers='[]'`.

## Help

```
WebRTC tunnel with Piping Server WebRTC signaling

Usage:
  webrtc-piping [flags]
  webrtc-piping [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  duplex      Duplex communication
  help        Help about any command
  tunnel      Tunneling TCP or UDP

Flags:
      --dns-server string    DNS server (e.g. 1.1.1.1:53)
  -H, --header stringArray   HTTP header
  -h, --help                 help for webrtc-piping
      --ice-servers json     ICE servers (default [{"urls":["stun:stun.l.google.com:19302"]}])
  -k, --insecure             Allow insecure server connections when using SSL
  -s, --server string        Piping Server URL (default "https://ppng.io")
  -v, --verbose              verbose output
  -V, --version              show version

Use "webrtc-piping [command] --help" for more information about a command.
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

## [0.5.0] - 2023-03-20
### Changed
* Improve `--ice-servers` default value in help

### Fixed
* Support `--insecure` and `--dns-server` options for "duplex" subcommand

### Added
* Add `-i` short option for `--ice-servers`

## [0.4.0] - 2023-03-19
### Changed
* Not show usage when failed
* Include SDP and candidates to log with -v

### Added
* Create `--ice-servers` option

## [0.3.0] - 2023-03-19
### Changed
* Rename webrtc-piping-tunnel → webrtc-piping
* Create "tunnel" and "duplex" subcommands
* Update dependencies

## [0.2.0] - 2022-07-04
### Added
* Support UDP port forwarding
* Add --insecure option
* Add --dns-server option
* Add --header option

## 0.1.0 - 2022-07-02
### Added
* Initial release

[Unreleased]: https://github.com/nwtgck/go-webrtc-piping/compare/v0.5.0...HEAD
[0.5.0]: https://github.com/nawtgck/go-webrtc-piping/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/nwtgck/go-webrtc-piping/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/nwtgck/go-webrtc-piping/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/nwtgck/go-webrtc-piping/compare/v0.1.0...v0.2.0


cmd/root.go
package cmd

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"github.com/nwtgck/go-webrtc-piping/version"
	"github.com/pion/webrtc/v3"
	"github.com/spf13/cobra"
	"net"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	ServerUrlEnvName = "PIPING_SERVER"
)

var flags struct {
	pipingServerUrl        string
	dnsServer              string
	insecure               bool
	httpHeaderKeyValueStrs []string
	iceServers             []iceServerFlag
	showsVersion           bool
	verbose                bool
}

type iceServerFlag struct {
	URLs       iceServerFlagUrls `json:"urls"`
	Username   string            `json:"username,omitempty"`
	Credential string            `json:"credential,omitempty"`
	// NOTE: credentialType is deprecated
}

// iceServerFlagUrls is string or []string
type iceServerFlagUrls struct {
	values []string
}

var _ json.Marshaler = (*iceServerFlagUrls)(nil)
var _ json.Unmarshaler = (*iceServerFlagUrls)(nil)

func (u *iceServerFlagUrls) MarshalJSON() ([]byte, error) {
	if len(u.values) == 1 {
		return json.Marshal(u.values[0])
	}
	return json.Marshal(u.values)
}

func (u *iceServerFlagUrls) UnmarshalJSON(b []byte) error {
	var stringValue string
	err := json.Unmarshal(b, &stringValue)
	if err == nil {
		u.values = []string{stringValue}
		return nil
	}
	err = json.Unmarshal(b, &u.values)
	if err == nil {
		return nil
	}
	return fmt.Errorf("urls is not string or string array")
}

func init() {
	cobra.OnInitialize()
	defaultServer, ok := os.LookupEnv(ServerUrlEnvName)
	if !ok {
		defaultServer = "https://ppng.io"
	}
	flags.iceServers = []iceServerFlag{
		{URLs: iceServerFlagUrls{values: []string{"stun:stun.l.google.com:19302"}}},
	}
	RootCmd.PersistentFlags().StringVarP(&flags.pipingServerUrl, "server", "s", defaultServer, "Piping Server URL")
	RootCmd.PersistentFlags().StringVar(&flags.dnsServer, "dns-server", "", "DNS server (e.g. 1.1.1.1:53)")
	// --insecure, -k is inspired by curl
	RootCmd.PersistentFlags().BoolVarP(&flags.insecure, "insecure", "k", false, "Allow insecure server connections when using SSL")
	RootCmd.PersistentFlags().StringArrayVarP(&flags.httpHeaderKeyValueStrs, "header", "H", []string{}, "HTTP header")
	RootCmd.PersistentFlags().VarP(&JSONFlag{Value: &flags.iceServers}, "ice-servers", "i", "ICE servers")
	RootCmd.PersistentFlags().BoolVarP(&flags.showsVersion, "version", "V", false, "show version")
	RootCmd.PersistentFlags().BoolVarP(&flags.verbose, "verbose", "v", false, "verbose output")
}

var RootCmd = &cobra.Command{
	Use:          os.Args[0],
	Short:        "webrtc-piping",
	Long:         "WebRTC tunnel with Piping Server WebRTC signaling",
	SilenceUsage: true,
	RunE: func(cmd *cobra.Command, args []string) error {
		if flags.showsVersion {
			fmt.Println(version.Version)
			return nil
		}
		return cmd.Help()
	},
}

func createHttpClient(insecureSkipVerify bool, dnsServer string /* empty string OK */) *http.Client {
	// Set insecure or not
	tr := &http.Transport{
		TLSClientConfig:   &tls.Config{InsecureSkipVerify: insecureSkipVerify},
		ForceAttemptHTTP2: true,
	}
	if dnsServer != "" {
		tr.DialContext = createDialContext(dnsServer)
	}
	return &http.Client{Transport: tr}
}

// Set default resolver for HTTP client
func createDialContext(dnsServer string) func(ctx context.Context, network, address string) (net.Conn, error) {
	resolver := &net.Resolver{
		PreferGo: true,
		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
			d := net.Dialer{
				Timeout: time.Millisecond * time.Duration(10000),
			}
			return d.DialContext(ctx, "udp", dnsServer)
		},
	}

	// Resolver for HTTP
	return func(ctx context.Context, network, address string) (net.Conn, error) {
		d := net.Dialer{
			Timeout:  time.Millisecond * time.Duration(10000),
			Resolver: resolver,
		}
		return d.DialContext(ctx, network, address)
	}
}

func parseHeaderKeyValueStrs(strKeyValues []string) ([][]string, error) {
	var keyValues [][]string
	for _, str := range strKeyValues {
		splitted := strings.SplitN(str, ":", 2)
		if len(splitted) != 2 {
			return nil, fmt.Errorf("invalid header format '%s'", str)
		}
		keyValues = append(keyValues, splitted)
	}
	return keyValues, nil
}

func createWebrtcConfig() webrtc.Configuration {
	iceServer := make([]webrtc.ICEServer, len(flags.iceServers))
	for i, d := range flags.iceServers {
		iceServer[i] = webrtc.ICEServer{
			URLs:       d.URLs.values,
			Username:   d.Username,
			Credential: d.Credential,
		}
	}
	return webrtc.Configuration{
		ICEServers: iceServer,
	}
}
